mdsuite/calculators/structure_factor.py

```diff
@@ -122,52 +125,71 @@ def __init__(self, **kwargs):
         self.y_label = r"$$\text{S(Q)}$$"
         self.analysis_name = "total_structure_factor"
 
+        self.result_series_keys = ["a", "b"]
+
         self.rho = None
 
         with open_text(data, "form_fac_coeffs.csv") as file:
             self.coeff_atomic_formfactor = pd.read_csv(
                 file, sep=","
             )  # stores coefficients for atomic form factors
 
-    def __call__(self, plot=True, data_range=1):
+    @call
+    def __call__(self, plot=True, save=True, data_range=1):
         """
         Parameters
         ----------
         plot : bool (default=True)
-                            Decision to plot the analysis.
+                                                Decision to plot the analysis.
         data_range : int (default=500)
-                            Range over which the property should be evaluated.
-                            This is not applicable to the current analysis as
-                            the full rdf will be calculated.
+                    Range over which the property should be evaluated.
+                    This is not applicable to the current analysis as
+                    the full rdf will be calculated.
 
         Returns
         -------
         None.
         """
 
+        self.args = Args(data_range=data_range)
+
         out = {}
-        for experiment in self.experiments:
-            self.experiment = experiment
 
-            self.plot = plot
+        self.number_of_atoms = 0
+        for species in self.experiment.species:
+            print("species", species)
+            self.number_of_atoms += self.experiment.species[species]["n_particles"]
 
-            self.rho = self.experiment.number_of_atoms / (
+        for species in self.experiment.species:
+            self.elements_dict[species] = {}
+            self.elements_dict[species]["particle_density"] = self.experiment.species[
+                species
+            ]["n_particles"] / (
                 self.experiment.box_array[0]
                 * self.experiment.box_array[1]
                 * self.experiment.box_array[2]
             )
 
-            if self.load_data:
-                out[self.experiment.name] = self.experiment.export_property_data(
-                    {"Analysis": self.analysis_name}
-                )
-            else:
-                out[self.experiment.name] = self.run_analysis()
+            self.elements_dict[species]["molar_fraction"] = (
+                self.experiment.species[species]["n_particles"] / self.number_of_atoms
+            )
```

Comment:
I will mention this latter, but I would prefer the previous version with `self.particle_density = {}` and `self.molar_fractions = {}` to avoid the nested dicts.

Reply:
same as before